shared/js/passcode_helper.js

```diff
@@ -0,0 +1,161 @@
+/* exported PasscodeHelper */
+/* globals crypto, TextEncoder */
+(function(exports) {
+  'use strict';
+
+  const SET_DIGEST_VALUE = 'lockscreen.passcode-lock.digest.value';
+  const SET_DIGEST_SALT = 'lockscreen.passcode-lock.digest.salt';
+  const SET_DIGEST_ITERATIONS = 'lockscreen.passcode-lock.digest.iterations';
+  const SET_DIGEST_ALGORITHM = 'lockscreen.passcode-lock.digest.algorithm';
+  const DEFAULT_ALGORITHM = 'SHA-1'; //XXX Update to SHA-256 after bug 554827?
+  const DEFAULT_ITERATIONS = 1000;
+
+
+  var PasscodeHelper = function() {
+    /* Usage:
+     *      var ph = new PasscodeHelper();
+     *      ph.setPassccode(string) -> Promise => digest (or false)
+     *      ph.checkPasscode(string) -> Promise resolves to a boolean
+     * */
+
+
+    var _salt;
+    var _iterations;
+    var _algorithm;
+    var _testPass;
+
+    this._encode = function(str) {
+      return new TextEncoder('utf-8').encode(str);
+    };
+    this._toTypedArray = function(obj) {
+      // SettingsAPI doesnt like arrays and gives us an Object { 1: .., 2: .. }
+      var a = [];
+      for (var key in obj) {
+        a.push(obj[key]);
+      }
+      return new Uint8Array(a);
+    };
+    this._deriveBits = function(pwKey, salt, iterations, algorithm) {
+      var length = 256;
```

Comment:
If we have the possibility to change the parameters of our algorithm, the name `_deriveBits` may need to be changed as some more specific name, like `derivePBKDF2Bits`. However I'm not security guru so I don't know the detail, so it depends on you and your design.